Source/Events.Processing/Filters/FiltersService.cs

```diff
@@ -123,6 +128,134 @@ public class FiltersService : FiltersBase
+            ServerCallContext context)
+        {
+            var dispatcher = _reverseCallDispatchers.GetFor<PartitionedFiltersClientToRuntimeMessage, FilterRuntimeToClientMessage, PartitionedFiltersRegistrationRequest, FilterRegistrationResponse, FilterEventRequest, PartitionedFilterResponse>(
+                runtimeStream,
+                clientStream,
+                context,
+                _ => _.RegistrationRequest,
+                (serverMessage, registrationResponse) => serverMessage.RegistrationResponse = registrationResponse,
+                (serverMessage, request) => serverMessage.FilterRequest = request,
+                _ => _.FilterResult,
+                _ => _.CallContext,
+                (request, context) => request.CallContext = context,
+                _ => _.CallContext);
+
+            if (!await dispatcher.ReceiveArguments(context.CancellationToken).ConfigureAwait(false))
+            {
+                const string message = "Filters connection arguments were not received";
+                _logger.Warning(message);
+                var failure = new Failure(FiltersFailures.NoFilterRegistrationReceived, message);
+                await PartitionedWriteFailedRegistrationResponse(dispatcher, failure, context.CancellationToken).ConfigureAwait(false);
+                return;
+            }
+
+            var arguments = dispatcher.Arguments;
+            var filterId = arguments.FilterId.To<StreamId>();
+            if (filterId.IsNonWriteable)
+            {
+                _logger.Warning("Received filter registration request with Filter Id: '{filterId}' which is an invalid stream id", filterId);
+                var failure = new Failure(FiltersFailures.CannotRegisterFilterOnNonWriteableStream, $"Received filter registration request with Filter Id: '{filterId}' which is an invalid stream id");
+                await PartitionedWriteFailedRegistrationResponse(dispatcher, failure, context.CancellationToken).ConfigureAwait(false);
+                return;
+            }
+
+            var scope = arguments.ScopeId.To<ScopeId>();
+            var streamId = StreamId.AllStreamId;
+
+            var registrationResults = await PartitionedRegisterStreamProcessorsForAllTenants(
+                scope,
+                StreamId.AllStreamId,
+                filterId,
+                dispatcher,
+                context.CancellationToken).ConfigureAwait(false);
+            try
+            {
+                if (!PartitionedTryStartStreamProcessors(registrationResults, out var failure))
+                {
+                    _logger.Warning(failure.Reason);
+                    await PartitionedWriteFailedRegistrationResponse(dispatcher, failure, context.CancellationToken).ConfigureAwait(false);
+                    return;
+                }
+
+                await dispatcher.Accept(new FilterRegistrationResponse(), context.CancellationToken).ConfigureAwait(false);
```

Comment:
It seems here like you start the StreamProcessors before the ReverseCall is accepted - which could then potentially start sending calls before the response is sent to the client - which is not allowed. Seems like there needs to be a two step registration here, where the Filters are created and registered (so they can't be started twice), and if that is succesfull, the ReverseCall should be accepted, and then the processing can commence.

Reply:
I'll register a task to ensure that the dispatcher is ready to receive calls before the stream prcessors is started